### PR TITLE
Draft remove runtime access

### PR DESCRIPTION
**Pull request by tgmichel**

### CODE REVIEW

File: rpc/src/eth.rs [hunk 1]
```diff
@@ -246,6 +253,39 @@ impl<B, C, P, CT, BE, A> EthApi<B, C, P, CT, BE, A> where
...
+
+	fn current_block(&self, id: &BlockId<B>) -> Option<ethereum::Block> {
+		if let Ok(Some(block_data)) = self.client.storage(
+			&id,
+			&StorageKey(
+				storage_prefix_build(b"Ethereum", b"CurrentBlock")
+			)
+		) {
+			return Some(Decode::decode(&mut &block_data.0[..]).unwrap());
+		} else { return None; };
```
Comment (JoshOrndorff) [inline rpc/src/eth.rs:265]: This is hard to read, and I'm not sure about the error handling. I think a match statement might helps. What does it mean when you get an `Err` vs an `Ok(None))`? Is it safe to unwrap here? If so, why not use `expect`?

File: rpc/src/eth.rs [hunk 2]
```diff
@@ -358,10 +398,8 @@ impl<B, C, P, CT, BE, A> EthApiT for EthApi<B, C, P, CT, BE, A> where
 			return Ok(None);
 		}
 
-		let block = self.client.runtime_api().current_block(&id)
-			.map_err(|err| internal_err(format!("call runtime failed: {:?}", err)))?;
-		let statuses = self.client.runtime_api().current_transaction_statuses(&id)
-			.map_err(|err| internal_err(format!("call runtime failed: {:?}", err)))?;
+		let block: Option<ethereum::Block> = self.current_block(&id);
+		let statuses: Option<Vec<TransactionStatus>> = self.current_statuses(&id);
```
Comment (JoshOrndorff) [inline rpc/src/eth.rs:402]: Why explicit type signatures?

Comment (JoshOrndorff) [reply rpc/src/eth.rs:402]: And why remove this error message? That could be really useful for someone who hits a bug in the future.

File: rpc/src/eth.rs [hunk 3]
```diff
@@ -804,12 +827,16 @@ impl<B, C, P, CT, BE, A> EthApiT for EthApi<B, C, P, CT, BE, A> where
 				return Ok(Vec::new());
 			}
 
-			let (block, _, statuses) = self.client.runtime_api()
-				.current_all(&id)
-				.map_err(|err| internal_err(format!("fetch runtime account basic failed: {:?}", err)))?;
+			let block: Option<ethereum::Block> = self.current_block(&id);
+			let statuses: Option<Vec<TransactionStatus>> = self.current_statuses(&id);
```
Comment (JoshOrndorff) [inline rpc/src/eth.rs:831]: You're replacing the call to `current_all` with individual calls. Doesn't that mean two calls into the runtime instead of one?

Comment (JoshOrndorff) [reply rpc/src/eth.rs:831]: Oh, or the entire point is that we're not using the runtime anymore. This is all calls to aux storage. If that's the case, why do we still have the stoage items in the runtime?

File: frame/ethereum/src/lib.rs [hunk 4]
```diff
@@ -279,8 +279,17 @@ impl<T: Trait> Module<T> {
...
+			CurrentReceipts::put(receipts.clone());
+		} else {
+			CurrentReceipts::kill();
+		}
+
+		if statuses.len() > 0 {
+			CurrentTransactionStatuses::put(statuses.clone());
+		} else {
+			CurrentTransactionStatuses::kill();
+		}
```
Comment (JoshOrndorff) [inline frame/ethereum/src/lib.rs:292]: How does this work? I thought these storage items should be cleared in `on_initialize`?

**Review by JoshOrndorff — COMMENTED** [review]
I've made some comments, but they probably just show my lack of familiarity with this code. I will need to understand it and the motiviation of this PR better before I can confidently approve it.

File: rpc/src/eth.rs (hunk 1 above)
Comment (tgmichel) [reply rpc/src/eth.rs:265]: > This is hard to read, and I'm not sure about the error handling. I think a match statement might helps.


I prefer to use if let when there are only two possible values to match.

> What does it mean when you get an Err vs an Ok(None))


Error can be https://crates.parity.io/sp_blockchain/enum.Error.html, and Ok(None) the storage value was not found for that StorageKey at that block height.

> Is it safe to unwrap here?


Yeah, Decode might return an error. Will fix!

File: rpc/src/eth.rs (hunk 2 above)
Comment (tgmichel) [reply rpc/src/eth.rs:402]: > Why explicit type signatures?

Readibility.

> And why remove this error message? That could be really useful for someone who hits a bug in the future.

We can add them!

File: rpc/src/eth.rs (hunk 3 above)
Comment (tgmichel) [reply rpc/src/eth.rs:831]: This are not auxiliary calls but calls to the StateBackend (we are accessing and decoding the storage primitives).

File: frame/ethereum/src/lib.rs (hunk 4 above)
Comment (tgmichel) [reply frame/ethereum/src/lib.rs:292]: > How does this work? I thought these storage items should be cleared in on_initialize?


So ugly isn't it? ^^ We need to move it somewhere else yes. Was just a way of get the thing working when I was doing the auxiliary logs stuff. 

Why is there: `receipts` and `statuses` vectors are initialized on top of the function and I want a `None` in those Storagevalues when there are no transactions in the current block. That way I can avoid inserting in them in auxiliary when I access this storagetdata from the frontier consensus.

**Comment by tgmichel** [comment]
@JoshOrndorff Your comments are very welcome and there are definitely a couple of draty things here, so thanks for the review!

The main problem we are trying to solve: RPC calls are free and makes no sense to invest lots of resources to compute them:

- Replace the runtime access for a merkle db search. This removes the runtime wasm spawn, the internal state apply and the search - that we'd be doing anyway when we access the storagevalue. Still not "free" but from our tests should be around 20 times faster (that's a lot, a request that in runtime takes i.e. 1 sec, through state db takes ~50millis.

- Additionally in this draft we store the transaction receipts logs in the auxiliary storage. The reason behind this is the same: we are trying to use the fastest access layer, but this time for smart contract "past events". We opted to not use state db for this, because the auxiliary is faster (around 20% faster than searching the merkle from our tests), and a single "past events" request may need to scan through 1000s of blocks (imagine using the runtime for that). **But**, still there is room for improvement in this - like storing the bloom filters in the header digest, scan matching those and then retrieve the log body from the auxiliary (although all of this still movies in my head, didn't test it!).

So to recap:

- We need to find a balance between storing the emulated block data in the substrate block state and quickly retrieving it.
- Make a moderate use of auxiliary as a helper for i.e. mapping (one-to-many block hash mapping) or deep scanning (past events).

File: rpc/src/eth.rs (hunk 3 above)
Comment (JoshOrndorff) [reply rpc/src/eth.rs:831]: Okay, so we're still reading the data stored in the runtime, but we're reading it in a more "direct" way by calling through the StateBacked, right?

Then my original question remains: Why replace the call to `current_all`? isn't it faster?

Comment (tgmichel) [reply rpc/src/eth.rs:831]: `current_all` belongs to the rpc runtime API impl. It's a single call to the runtime that loads 3 storage values from pallet-ethereum, but still a runtime call.
